xwiki-pro-macros-api/src/main/java/com/xwiki/macros/excerptinclude/internal/listener/ExcerptIncludeMacroListener.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.macros.excerptinclude.internal.listener;
+
+import java.util.Arrays;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.hibernate.Session;
+import org.slf4j.Logger;
+import org.xwiki.bridge.event.DocumentCreatedEvent;
+import org.xwiki.bridge.event.DocumentDeletedEvent;
+import org.xwiki.bridge.event.DocumentUpdatedEvent;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.event.Event;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.block.match.MacroBlockMatcher;
+import org.xwiki.stability.Unstable;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.store.XWikiHibernateBaseStore.HibernateCallback;
+import com.xwiki.macros.excerptinclude.internal.StoreHandler;
+
+/**
+ * Listens to created, updated or deleted pages, checks for Excerpt Include Macro and adds/removes backlinks for the
+ * excerpt macro page referenced in the excerpt-include macro. The backlinks are used to update the "0" parameter of the
+ * excerpt-include macro in case of excerpt page rename.
+ *
+ * @version $Id$
+ * @since 1.14.4
+ */
+@Component
+@Named(ExcerptIncludeMacroListener.ROLE_HINT)
+@Singleton
+@Unstable
+public class ExcerptIncludeMacroListener extends AbstractEventListener
+{
+    protected static final String ROLE_HINT = "ExcerptIncludeMacroListener";
+
+    @Inject
+    @Named("explicit")
+    private DocumentReferenceResolver<String> explicitDocumentReferenceResolver;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private StoreHandler storeHandler;
+
+    /**
+     * Constructor.
+     */
+    public ExcerptIncludeMacroListener()
+    {
+        super(ROLE_HINT, Arrays.<Event>asList(new DocumentCreatedEvent(), new DocumentDeletedEvent(),
+            new DocumentUpdatedEvent()));
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        XWikiDocument document = (XWikiDocument) source;
+        XWikiContext context = (XWikiContext) data;
+
+        List<Block> macroBlocks =
+            document.getXDOM().getBlocks(new MacroBlockMatcher("excerpt-include"), Block.Axes.DESCENDANT);
+
+        if (!macroBlocks.isEmpty()) {
+            try {
+                // We need to delete existing links before saving the page's ones.
+                storeHandler.deleteLinks(document.getId(), context);
```

Comment:
Is it really this extension's job to delete all links for the document? Doesn't this also delete links that won't be re-added here? Also, why are links only deleted when there are macros - shouldn't they also be removed when there are no macros in case there was a macro before updating the document or when the document is deleted?